addons/controllers/cpi/vspherecpiconfig_utils.go

```diff
@@ -20,13 +20,58 @@ import (
 	capvv1beta1 "sigs.k8s.io/cluster-api-provider-vsphere/apis/v1beta1"
 	capvvmwarev1beta1 "sigs.k8s.io/cluster-api-provider-vsphere/apis/vmware/v1beta1"
 	clusterapiv1beta1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	clusterapiutil "sigs.k8s.io/cluster-api/util"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	"github.com/vmware-tanzu/tanzu-framework/addons/pkg/constants"
 	pkgtypes "github.com/vmware-tanzu/tanzu-framework/addons/pkg/types"
 	"github.com/vmware-tanzu/tanzu-framework/addons/pkg/util"
 	cpiv1alpha1 "github.com/vmware-tanzu/tanzu-framework/apis/addonconfigs/cpi/v1alpha1"
 )
 
+// ClusterToVSphereCPIConfig returns a list of Requests with VSphereCPIConfig ObjectKey based on Cluster events
```

Comment:
I think the intent here is that for customisations, you can't use the templated vSphereCPIConfig.

However, this behaviour also seems bad. It looks like we're introducing explicit ordering into a system that isn't designed for it.

We need an option in a Cluster to explicitly *skip* vSphereCPIConfig templating otherwise the behaviour here is non-deterministic.

Reply:
Seems like we are covering this specific use case in https://github.com/vmware-tanzu/tanzu-framework/pull/3318#discussion_r969046352 and don't need to enqueue for Cluster events. 

If we really want to enqueue then VsphereCluster is a better object to use.

Reply:
@randomvariable @vijaykatam 
We decide to go on ordering route when we wants to infer the addon parameters from Cluster object. All addons relies on Cluster.Name to create the addon secret. For example Antrea https://github.com/vmware-tanzu/tanzu-framework/blob/main/addons/controllers/antrea/antreaconfig_controller.go#L168

At the same time, currently all addonController will exit silently if the owner Cluster is not found.
Antrea https://github.com/vmware-tanzu/tanzu-framework/blob/main/addons/controllers/antrea/antreaconfig_controller.go#L83
Calico https://github.com/vmware-tanzu/tanzu-framework/blob/main/addons/controllers/calico/calicoconfig_controller.go#L80
CSI https://github.com/vmware-tanzu/tanzu-framework/blob/main/addons/controllers/csi/vspherecsiconfig_utils.go#L180
KAPP https://github.com/vmware-tanzu/tanzu-framework/blob/main/addons/controllers/kapp-controller/kappcontrollerconfig_controller.go#L77
Since we only observe delay with CPI and now we only deal with CPI. But for other addon there could be potentially same issues as well.
KAPP controller already enqueue events of cluster https://github.com/vmware-tanzu/tanzu-framework/blob/main/addons/controllers/kapp-controller/kappcontrollerconfig_controller.go#L97-L100

Reply:
But enqueue could cause race condition looks like
```
==================
WARNING: DATA RACE
Write at 0x00c0014dfe50 by goroutine 134:
  github.com/vmware-tanzu/tanzu-framework/addons/controllers/cpi.(*VSphereCPIConfigReconciler).Reconcile()
      /home/runner/work/tanzu-framework/tanzu-framework/addons/controllers/cpi/vspherecpiconfig_controller.go:84 +0x1cf
  sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile()
      /home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:114 +0x39b
  sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler()
      /home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:311 +0x43a
  sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem()
      /home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:266 +0x350
  sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
      /home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:227 +0xd7

Previous read at 0x00c0014dfe50 by goroutine 18:
  runtime.racereadrange()
      <autogenerated>:1 +0x1b
  github.com/vmware-tanzu/tanzu-framework/addons/controllers/cpi.(*VSphereCPIConfigReconciler).VSphereClusterToVSphereCPIConfig-fm()
      /home/runner/work/tanzu-framework/tanzu-framework/addons/controllers/cpi/vspherecpiconfig_utils.go:34 +0x4d
  sigs.k8s.io/controller-runtime/pkg/handler.(*enqueueRequestsFromMapFunc).mapAndEnqueue()
      /home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/handler/enqueue_mapped.go:80 +0x84
  sigs.k8s.io/controller-runtime/pkg/handler.(*enqueueRequestsFromMapFunc).Create()
      /home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/handler/enqueue_mapped.go:57 +0xd7
  sigs.k8s.io/controller-runtime/pkg/source/internal.EventHandler.OnAdd()
      /home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/source/internal/eventsource.go:63 +0x311
  sigs.k8s.io/controller-runtime/pkg/source/internal.(*EventHandler).OnAdd()
      <autogenerated>:1 +0xda
  k8s.io/client-go/tools/cache.(*processorListener).run.func1()
      /home/runner/go/pkg/mod/k8s.io/client-go@v0.23.5/tools/cache/shared_informer.go:787 +0x12e
  k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1()
      /home/runner/go/pkg/mod/k8s.io/apimachinery@v0.23.5/pkg/util/wait/wait.go:155 +0x81
  k8s.io/apimachinery/pkg/util/wait.BackoffUntil()
      /home/runner/go/pkg/mod/k8s.io/apimachinery@v0.23.5/pkg/util/wait/wait.go:156 +0xce
  k8s.io/apimachinery/pkg/util/wait.JitterUntil()
      /home/runner/go/pkg/mod/k8s.io/apimachinery@v0.23.5/pkg/util/wait/wait.go:133 +0x104
  k8s.io/apimachinery/pkg/util/wait.Until()
      /home/runner/go/pkg/mod/k8s.io/apimachinery@v0.23.5/pkg/util/wait/wait.go:90 +0x78
  k8s.io/client-go/tools/cache.(*processorListener).run()
      /home/runner/go/pkg/mod/k8s.io/client-go@v0.23.5/tools/cache/shared_informer.go:781 +0x18
  k8s.io/client-go/tools/cache.(*processorListener).run-fm()
      /home/runner/go/pkg/mod/k8s.io/client-go@v0.23.5/tools/cache/shared_informer.go:775 +0x39
  k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1()
      /home/runner/go/pkg/mod/k8s.io/apimachinery@v0.23.5/pkg/util/wait/wait.go:73 +0x7e
```

Reply:
@lubronzhan Per my understanding on controller runtime, if VSphereCPIConfigReconciler returns an error when a VSphereCPIConfig's owner Cluster is not found, the reconcile request will be requeued repeatedly untill the owner Cluster is created. Why is it `reequeue 20 minutes later`? I am not sure if it is required to watch Cluster events.

BTW, no matter we enqueue for Cluster or vSphereCluster events, the vspherecpiconfig controller also catch non-classy cluster events which seem to be not efficient.

Reply:
>if VSphereCPIConfigReconciler returns an error when a VSphereCPIConfig's owner Cluster is not found, the reconcile request will be requeued repeatedly untill the owner Cluster is created. Why is it reequeue 20 minutes later?
I was talking about if there is no error returned, which is only fixed inside this PR. Then it will take 20 minutes to enqueue. I still don't like the idea of returning error when Cluster is not found. Since customer could create some Addon Config themselves without referring to any cluster.


> the vspherecpiconfig controller also catch non-classy cluster events which seem not to be efficient.
It's good practice for controller to watch on Resources that the API is replying on. You can see Kapp do the same, also CAPI https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/main/controllers/vspherecluster_controller.go#L137-L156